lib/dice/src/handoff.rs

```diff
@@ -0,0 +1,90 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::{AliasCert, AliasOkm, DeviceIdCert};
+use hubpack::SerializedSize;
+use lpc55_pac::syscon::RegisterBlock;
+use serde::{Deserialize, Serialize};
+
+// This memory is the USB peripheral SRAM that's 0x4000 bytes long. Changes
+// to this address must be coordinated with the [dice_*] tables in
+// chips/lpc55/chip.toml
+// TODO: get from app.toml -> chip.toml at build time
+const MEM_START: usize = 0x4010_0000;
+const ALIAS_START: usize = MEM_START;
+
+/// The Handoff type is a thin wrapper over the memory region used to transfer
+/// DICE artifacts (seeds & certs) from stage0 to hubris tasks. It is intended
+/// for use by stage0 to write these artifacts to memory where they will later
+/// be read out by a hubris task.
+pub struct Handoff<'a>(&'a RegisterBlock);
+
+impl<'a> Handoff<'a> {
+    pub fn turn_on(syscon: &'a RegisterBlock) -> Self {
+        // handoff through USB SRAM requires we power it on
+        syscon.ahbclkctrl2.modify(|_, w| w.usb1_ram().enable());
+        syscon
+            .presetctrl2
+            .modify(|_, w| w.usb1_ram_rst().released());
+
+        Self(syscon)
+    }
+
+    pub fn turn_off(self) {
+        self.0
+            .presetctrl2
+            .modify(|_, w| w.usb1_ram_rst().asserted());
+        self.0.ahbclkctrl2.modify(|_, w| w.usb1_ram().disable());
+    }
+
+    pub fn alias(&self, alias_data: &AliasData) -> usize {
+        let dst_ptr = ALIAS_START as *mut [u8; AliasData::MAX_SIZE];
+        // SAFETY: Dereferencing this raw pointer is necessary to write to the
+        // memory region used to handoff DICE artifacts to Hubris tasks. This
+        // pointer will references a valid memory region provided two
+        // conditions are met:
+        // 1) The associated memory region has been enabled / turned on if
+        // necessary. This happens in the constructor / 'turn_on' function.
+        // 2) The function call is made by code in stage0.
```

Comment:
Accessing the SRAM when its powered off will land you in a fault handler. We could insure against this situation by enabling the memory unconditionally in stage0. It could still be disabled further down the line but this is partially mitigated by restricting access to the syscon driver. After this the only other condition is sufficient permissions to access the memory region which means either being in stage0 (where this interface should be used) or having been granted access by the hubris kernel. Whether or not the whole method becomes unsafe because of this is rightly a question for those you've tagged above.

Reply:
I don't think that that means this becomes unsafe. The key question with unsafe is, "If I call this method with a certain parameter, will I introduce memory unsafety?" If the answer is, in all cases, that you end up in a fault handler, doesn't sound like memory unsafety to me.

Reply:
Feels worth clarifying in the safety comment that this is safe even if these conditions aren't met since it will cause a fault.